rules/recording.go

```diff
@@ -77,14 +79,114 @@ func (rule *RecordingRule) Labels() labels.Labels {
 	return rule.labels
 }
 
+func (rule *RecordingRule) EvalWithExemplars(ctx context.Context, queryOffset time.Duration, ts time.Time, query QueryFunc,
+	exemplarQuery ExemplarQueryFunc, _ *url.URL, limit int,
+) (promql.Vector, []exemplar.QueryResult, error) {
+	ctx = NewOriginContext(ctx, NewRuleDetail(rule))
+	vector, err := rule.eval(ctx, ts.Add(-queryOffset), query)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	selectors := parser.ExtractSelectors(rule.vector)
```

Comment:
Maybe there needs to be a special case for `__name__`? Because I think otherwise this immediately disqualifies all recording rules what use stuff like `rate()` (very common)

Reply:
Let me get back to you once I validated behavior as part of additional tests.

Reply:
I added a test `TestComplexRuleEvalWithExemplars` for expression
```
  sum without (label_a) (rate(request_failures[2m]))
/
  sum without (label_a) (rate(requests[2m]))
```

For `rate` and other functions, the underlying metric name can be used just like in the other examples.